release/context.go

```diff
@@ -76,90 +86,86 @@ func (rc *ReleaseContext) Clean() {
-	// be selected for upgrading/applying.
+// files and the running platform.
+// ServiceFilter's can be provided to filter the found services.
+// Be careful about the ordering of the filters. Filters that are earlier
+// in the slice will have higher priority (they are run first).
+func (rc *ReleaseContext) SelectServices(results flux.ReleaseResult, logStatus statusFn, filters ...ServiceFilter) ([]*ServiceUpdate, error) {
+	// Get services defined in repository
 	defined, err := rc.FindDefinedServices()
 	if err != nil {
 		return nil, err
 	}
 
-	updateMap := map[flux.ServiceID]*ServiceUpdate{}
 	var ids []flux.ServiceID
-
-	// The services to explicitly include; this is left as nil if only
-	// was not given, since we treat missing services slightly
-	// differently depending on whether an exact set was requested.
-	var only flux.ServiceIDSet
-	if included != nil {
-		only = flux.ServiceIDSet{}
-		only.Add(included)
-	}
-
+	definedMap := map[flux.ServiceID]*ServiceUpdate{}
 	for _, s := range defined {
-		if only == nil || only.Contains(s.ServiceID) {
-			var result flux.ServiceResult
-			switch {
-			case excluded.Contains(s.ServiceID):
-				logStatus("Skipping service %s as it is excluded", s.ServiceID)
-				result = flux.ServiceResult{
-					Status: flux.ReleaseStatusSkipped,
-					Error:  "excluded",
-				}
-			case locked.Contains(s.ServiceID):
-				logStatus("Skipping service %s as it is locked", s.ServiceID)
-				result = flux.ServiceResult{
-					Status: flux.ReleaseStatusSkipped,
-					Error:  "locked",
-				}
-			default:
-				result = flux.ServiceResult{
-					Status: flux.ReleaseStatusPending,
-				}
-				updateMap[s.ServiceID] = s
-				ids = append(ids, s.ServiceID)
-			}
-			results[s.ServiceID] = result
-		}
+		ids = append(ids, s.ServiceID)
+		definedMap[s.ServiceID] = s
 	}
 
-	// Correlate with services in running system.
+	// Get running services from cluster
 	services, err := rc.Instance.GetServices(ids)
 	if err != nil {
 		return nil, err
 	}
 
+	// Compare defined vs running
 	var updates []*ServiceUpdate
-	for _, service := range services {
-		logStatus("Found service %s", service.ID)
-		update := updateMap[service.ID]
-		update.Service = service
+	for _, s := range services {
+		logStatus("Found service %s", s.ID)
+		update := definedMap[s.ID]
+		update.Service = s
 		updates = append(updates, update)
-		delete(updateMap, service.ID)
+		delete(definedMap, s.ID)
 	}
-	// Mark anything left over as skipped
-	for id, _ := range updateMap {
-		ignoringOrSkipping := "Ignoring"
-		status := flux.ReleaseStatusIgnored
-		if included != nil { // i.e., this service was specifically requested
-			ignoringOrSkipping = "Skipping"
-			status = flux.ReleaseStatusSkipped
+
+	// Filter both updates ...
+	var filteredUpdates []*ServiceUpdate
+	for _, s := range updates {
+		fr := s.filter(filters...)
+		if fr.Result.Error != "" {
+			logStatus(fr.String())
+		}
+		results[s.ServiceID] = fr.Result
```

Comment:
Yes, there's always a `Result`. A `Result` is the thing that is shown to the user. This is separate from a `ServiceUpdate` that actually does the updating.

All defined services have a result. The ones that have the status type "ignored" will be ignored by the Printer unless verbose is selected: https://github.com/weaveworks/flux/blob/master/release/print.go#L19